plugins/historydata/ua_history_data_backend_memory.c

```diff
@@ -603,3 +606,268 @@ UA_HistoryDataBackend_Memory_clear(UA_HistoryDataBackend *backend)
+
+    if(item->dataStore[item->lastInserted] != NULL) {
+        UA_DataValueMemoryStoreItem_clear(item->dataStore[item->lastInserted]);
+        UA_free(item->dataStore[item->lastInserted]);
+    }
+    item->dataStore[item->lastInserted] = newItem;
+    ++item->lastInserted;
+
+    if(item->storeEnd < item->storeSize) {
+        ++item->storeEnd;
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static size_t
+getResultSize_service_Circular(const UA_HistoryDataBackend *backend, UA_Server *server,
+                               const UA_NodeId *sessionId, void *sessionContext,
+                               const UA_NodeId *nodeId, UA_DateTime start,
+                               UA_DateTime end, UA_UInt32 numValuesPerNode,
+                               UA_Boolean returnBounds, size_t *startIndex,
+                               size_t *endIndex, UA_Boolean *addFirst,
+                               UA_Boolean *addLast, UA_Boolean *reverse) {
+    *startIndex = 0;
+    *endIndex =
+        backend->lastIndex(server, backend->context, sessionId, sessionContext, nodeId);
+    *addFirst = false;
+    *addLast = false;
+    if(end == LLONG_MIN) {
+        *reverse = false;
+    } else if(start == LLONG_MIN) {
+        *reverse = true;
+    } else {
+        *reverse = end < start;
+    }
+
+    size_t size = 0;
+    const UA_NodeIdStoreContextItem_backend_memory *item =
+        getNodeIdStoreContextItem_backend_memory_Circular(
+            (UA_MemoryStoreContext *)backend->context, server, nodeId);
+    if(item == NULL) {
+        size = 0;
+    } else {
+        size = item->storeEnd;
+    }
+    return size;
+}
+
+static UA_StatusCode
+getHistoryData_service_Circular(
+    UA_Server *server, const UA_NodeId *sessionId, void *sessionContext,
+    const UA_HistoryDataBackend *backend, const UA_DateTime start, const UA_DateTime end,
+    const UA_NodeId *nodeId, size_t maxSize, UA_UInt32 numValuesPerNode,
+    UA_Boolean returnBounds, UA_TimestampsToReturn timestampsToReturn,
+    UA_NumericRange range, UA_Boolean releaseContinuationPoints,
+    const UA_ByteString *continuationPoint, UA_ByteString *outContinuationPoint,
+    UA_HistoryData *historyData) {
+    size_t *resultSize = &historyData->dataValuesSize;
+    UA_DataValue **result = &historyData->dataValues;
+    size_t skip = 0;
+    UA_ByteString backendContinuationPoint;
+    UA_ByteString_init(&backendContinuationPoint);
+    if(continuationPoint->length > 0) {
+        if(continuationPoint->length >= sizeof(size_t)) {
+            skip = *((size_t *)(continuationPoint->data));
+            if(continuationPoint->length > 0) {
+                backendContinuationPoint.length =
+                    continuationPoint->length - sizeof(size_t);
+                backendContinuationPoint.data = continuationPoint->data + sizeof(size_t);
+            }
+        } else {
+            return UA_STATUSCODE_BADCONTINUATIONPOINTINVALID;
```

Comment:
Treat the error case first.
Then you don't have to indent into an else condition.
Flat code is good.